Samples/NetworkMediaStreamer/RTSPServer/src/RtspSession.cpp

```diff
@@ -0,0 +1,575 @@
+        if (!m_dest.empty())
+            logstring += m_dest;
+        else
+            logstring += "tcp://" + m_RtspClientAddr;
+
+        m_pLoggerEvents[(int)LoggerType::OTHER](S_OK, winrt::to_hstring(logstring));
+    }
+    else
+    {
+        Response = "RTSP/1.0 401 Unauthorized\r\n"
+        + std::string("CSeq: ") + m_CSeq + "\r\n";
+
+        if (m_spAuthProvider)
+        {
+            winrt::hstring curAuthSessionMsg;
+            HSTRING msg;
+            winrt::check_hresult(m_spAuthProvider->GetNewAuthSessionMessage(&msg));
+            winrt::attach_abi(curAuthSessionMsg, msg);
+            m_curAuthSessionMsg = winrt::to_string(curAuthSessionMsg);
+            Response += m_curAuthSessionMsg;
+        }
+        Response += std::string("Server: NightKing\r\n")
+            + DateHeader() + "\r\n\r\n";
+
+        SendToClient(Response);
+    }
+
+    m_pLoggerEvents[(int)LoggerType::RTSPMSGS](S_OK, winrt::to_hstring(__FUNCTION__) + L":Response:" + winrt::to_hstring(Response));
+}
+
+void RTSPSession::HandleCmdTEARDOWN()
+{
+    StopIfStreaming();
+}
+
+void RTSPSession::Handle_RtspPAUSE()
+{
+    std::string   Response;// [1024] ;
+    Response = "RTSP/1.0 200 OK\r\nCSeq:" + m_CSeq + "\r\n"
+        + DateHeader() + "\r\n"
+        + "Range: npt=0.000-\r\n"
+        + "Session: " + std::to_string(m_RtspSessionID) + "\r\n"
+        + "RTP-Info: url=" + m_URLProto + "://" + m_URLHostPort + "\r\n\r\n",
+
+        SendToClient(Response);
+    StopIfStreaming();
+
+    m_pLoggerEvents[(int)LoggerType::RTSPMSGS](S_OK, winrt::to_hstring(__FUNCTION__) + L":Response:" + winrt::to_hstring(Response));
+
+}
+
+void RTSPSession::SendToClient(std::string Response)
+{
+    m_pRtspClient->Send((BYTE*)Response.c_str(), (int)Response.length());
+}
+
+char const* RTSPSession::DateHeader()
+{
+    static char buf[RTSP_PARAM_STRING_MAX];
+    time_t tt = time(NULL);
+    tm t;
+    gmtime_s(&t, &tt);
+    strftime(buf, sizeof buf, "Date: %a, %b %d %Y %H:%M:%S GMT", &t);
+    return buf;
+}
+
+int RTSPSession::GetStreamID()
+{
+    return 0;//m_StreamID;
+}
+
+void RTSPSession::BeginSession(winrt::delegate<RTSPSession*> completed)
+{
+    m_Completed = completed;
+    m_RtspReadEvent.attach(WSACreateEvent());      // create READ wait event for our RTSP client socket
+    WSAEventSelect(m_pRtspClient->GetSocket(), m_RtspReadEvent.get(), FD_READ | FD_CLOSE);   // select socket read event
+    RegisterWaitForSingleObject(m_callBackHandle.put(), m_RtspReadEvent.get(), [](PVOID arg, BOOLEAN flag)
+        {
+            auto pSession = (RTSPSession*)arg;
+            try
+            {
+                WSAResetEvent(pSession->m_RtspReadEvent.get());
+                auto l = std::lock_guard(pSession->m_readDelegateMutex);
+                if (!pSession->m_callBackHandle)
+                {
+                    //Session stopped
+                    return;
+                }
+
+                char* pRecvBuf = (char*)pSession->m_pTcpRxBuff.get();

```

Comment:
>m_pTcpRxBuff [](start = 50, length = 12)

is this ever not nullptr? I don't understand where this is assigned to something else than the default value of nullptr

Reply:
line 42 : in the constructor-
  m_pTcpRxBuff = std::make_unique<BYTE[]>(RTSP_BUFFER_SIZE);

---
In reply to: [492996618](https://github.com/microsoft/Windows-Camera/pull/13#discussion_r492996618) [](ancestors = 492996618)

Reply:
ok but then this is used to HandleRequest(), where is it assigned to something else?

Reply:
i do not understand this comment/question. This buffer is allocated in the constructor. It is used in the Recv call to receive the message/request from client and then handled/parsed.

---
In reply to: [493031118](https://github.com/microsoft/Windows-Camera/pull/13#discussion_r493031118) [](ancestors = 493031118)